### PR TITLE
[RFC] Factories composable relations

### PR DESCRIPTION
**Pull request by defudef**
- [x] I Accept [Individual Contributor License Agreement](https://github.com/DivanteLtd/next/blob/master/CLA.md)

### CODE REVIEW

File: packages/core/rfcs/composable-relations.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+```TS
+// useUserAddressFactory
+
+interface UseUserAddressFactoryParams {
+  userComposable?: UseUser<USER, USER_UPDATE_PARAMS>;
+  addAddress: { userUpdated: USER, addressesUpdated: ADDRESS[] }
+}
+
+export function useUserAddressFactory {
```
Comment (andrzejewsky) [inline packages/core/rfcs/composable-relations.md:66]: that's the weird example it's not even a factory but looks like factory params

**Comment by filrak** [comment]
The intention was very similar. The fact that `useuser` is calling the composable directly is our fault - we missed that during CR somehow.

The only change I would suggest is to provide explicitly a `dependencies` object into params that will be holding all external dependencies. Also i wouldn't suggest passing whole compostable there, rather a specific objects or functions because if we want to modify for example `cart` from `useCart` we can't do that from within comparable. Btw check `useCart` for reference, I suggest something similar so for example in `useUser` we will pass
```js
dependencies: {
    cart: cart
    refreshCart: useCart().refreshCart()
}
```

I really don't get that part - can you elaborate on what was the intention?

```js
export function useUserAddressFactory {
  return {
    userComposable: useUser(),
    addAddress: (address) => {
      const { userUpdated, adressessUpdated } = await factoryParams.addAddress(address);
      addresses.value = addresses;
      userComposable && userComposable.updateUser(userUpdated);
    }
  };
}
```

Imho factory should return only a comparable, whats the point of returning other comparable from this factory along with some methods? I'm suspecting you just made some mistake here and you meant factory params here?

File: packages/core/rfcs/composable-relations.md (hunk 1 above)
Comment (andrzejewsky) [reply packages/core/rfcs/composable-relations.md:66]: I was thinking about this DI etc. and it ONLY make sense when you need to call other composable INSIDE of the factory, but this way is a little bit risky as some of these calls can depend on the integration. Using factory params you can do something like this

```
const someComposable = useSomeComposable();

const factoryParams = {
  someFunction: async () => {
    const field = someComposable.someField;
  }
}
```

Comment (andrzejewsky) [reply packages/core/rfcs/composable-relations.md:66]: @defudef so please provide other examples where it won't work and you need DI mechanism

Comment (filrak) [reply packages/core/rfcs/composable-relations.md:66]: I'm actually against calling it that way tbh, we should have a way of performing such stuff from the inside of the factory so the control is on our side, O believe this is the only reason why we need such thing. Ideally we should have some kind of internal linking between compostables that needs to be linked (i believe it applies only to user

**Comment by filrak** [comment]
I just came up with another idea, probably much more ellegant. Solution proposed in this RFC is a bit overcomplicated and relies on the factory params in terms of injecting the data - would be nice if we could hide this relation inside factories. 

We can do tha from the inside of the factories. we just need an object that will work like a "mediator" between user cart and user subcomposables.

Mediator object could look like this
```js
const shared = {
  cart: ref(null),
  user: ref(null)
}
```

 Then we're just reading from this mediator object inside compostables that need it. 
```js
import { shared } from './shared'

function useUserAddressFactory() {
  const user = shared.user
  return function useUserAddresses() { ... }
}
```
And overriding if that's needed

```js
import { shared } from './shared'

// from useCartFactory
function useCartFactory () {
  const cart = shared.cart

  // later in code
  loadCart () {
     const updatedCart = factoryParams.loadCart()
     cart.value = updatedCart
  }
```

With that we can omit forcing user to directly inject the properties into factories which brings back the control to us and reduces complexity of the integration.

Another advantage would be keeping full treeshaking - previous approach ends up in grouping shared compostable in the same chunks

File: packages/core/rfcs/composable-relations.md (hunk 1 above)
Comment (defudef) [reply packages/core/rfcs/composable-relations.md:66]: Sorry, just missed brackets:

```TS
export function useUserAddressFactory() {
 // that stuff inside here
}
````

**Comment by filrak** [comment]
Closing. There is a new RFC on that topic https://github.com/DivanteLtd/next/pull/342

**Pull request closed without merging**